btc/src/main/kotlin/provider/btc/generation/BtcPublicKeyProvider.kt

```diff
@@ -72,48 +73,56 @@ class BtcPublicKeyProvider(
+            } else if (notaryKeys.size != peers) {
+                logger.info { "Not enough keys are collected to generate a multisig address(${notaryKeys.size} out of ${peers})" }
+                return@of
+            } else if (!hasMyKey(notaryKeys)) {
+                logger.info { "Cannot be involved in address generation. No access to $notaryKeys." }
+                return@of
+            }
+            val threshold = getSignThreshold(peers)
+            val msAddress = createMsAddress(notaryKeys, threshold)
+            if (!walletFile.wallet.addWatchedAddress(msAddress)) {
+                throw IllegalStateException("BTC address $msAddress was not added to wallet")
+            }
+            logger.info { "Address $msAddress was added to wallet." }
 
-                val (addressInfo, storageAccount) = when (addressType) {
-                    BtcAddressType.CHANGE -> {
-                        logger.info { "Creating change address" }
-                        Pair(
-                            AddressInfo.createChangeAddressInfo(ArrayList<String>(notaryKeys)),
-                            changeAddressStorageAccount
-                        )
-                    }
-                    BtcAddressType.FREE -> {
-                        logger.info { "Creating free address" }
-                        Pair(AddressInfo.createFreeAddressInfo(ArrayList<String>(notaryKeys)), notaryAccount)
-                    }
+            val (addressInfo, storageAccount) = when (addressType) {
+                BtcAddressType.CHANGE -> {
+                    logger.info { "Creating change address" }
+                    Pair(
+                        AddressInfo.createChangeAddressInfo(ArrayList<String>(notaryKeys)),
+                        changeAddressStorageAccount
+                    )
+                }
+                BtcAddressType.FREE -> {
+                    logger.info { "Creating free address" }
+                    Pair(AddressInfo.createFreeAddressInfo(ArrayList<String>(notaryKeys)), notaryAccount)
                 }
-                ModelUtil.setAccountDetail(
-                    multiSigConsumer,
-                    storageAccount,
-                    msAddress.toBase58(),
-                    addressInfo.toJson()
-                ).fold({
-                    //TODO this save will probably corrupt the wallet file
-                    walletFile.save()
-                    logger.info { "New BTC ${addressType.title} address $msAddress was created " }
-                }, { ex -> throw ex })
             }
+            ModelUtil.setAccountDetail(
+                multiSigConsumer,
+                storageAccount,
+                msAddress.toBase58(),
+                addressInfo.toJson(),
+                BigInteger.valueOf(generationTime)
+            ).fold({
+                //TODO this save will probably corrupt the wallet file
```

Comment:
Is it bad? What should we do with that in the future?

Reply:
It's very bad. But it also outdated)